graph/pull_v2.go

```diff
@@ -232,60 +234,101 @@ func (p *v2Puller) pullV2Tag(tag, taggedName string) (bool, error) {
 
 		p.graph.Retain(p.sessionID, img.ID)
 		layerIDs = append(layerIDs, img.ID)
+	}
+
+	// Use this separate loop to launch the goroutines which will perform the
+	// downloads in parallel. We need a separate loop here *without* returns so
+	// that we don't launch some goroutines and then exit before they all
+	// terminate.
+	for i := len(manifest.FSLayers) - 1; i >= 0; i-- {
+		download := &downloads[i] // Note: we need a pointer to the download.
+		img := download.img
 
 		// Check if exists
 		if p.graph.Exists(img.ID) {
 			logrus.Debugf("Image already exists: %s", img.ID)
+			out.Write(p.sf.FormatProgress(stringid.TruncateID(img.ID), "Already exists", nil))
 			continue
 		}
 
 		out.Write(p.sf.FormatProgress(stringid.TruncateID(img.ID), "Pulling fs layer", nil))
 
-		downloads[i].err = make(chan error)
-		go p.download(&downloads[i])
+		download.err = make(chan error)
+		go p.download(download)
 	}
 
-	var tagUpdated bool
+	// NOTE: We cannot return *at all* within this loop. We must wait for every
+	// download goroutine to terminate before returning. When we reach the end
+	// of this loop, and pullErred is true, then you have my permission to die.
+	var tagUpdated, pullErred bool
 	for i := len(downloads) - 1; i >= 0; i-- {
 		d := &downloads[i]
-		if d.err != nil {
-			if err := <-d.err; err != nil {
-				return false, err
-			}
+
+		if d.err == nil {
```

Comment:
this is not your fault but `d.err` doesn't sound like a channel, it feels like `err == nil` means no errors.